pkg/apis/batch/types.go

```diff
@@ -183,9 +183,11 @@ type JobSpec struct {
 	// for each index.
 	// When value is `Indexed`, .spec.completions must be specified and
 	// `.spec.parallelism` must be less than or equal to 10^5.
+	// In addition, The Pod name takes the form
+	// `$(job-name)-$(index)-$(random-string)`,
+	// the Pod hostname takes the form `$(job-name)-$(index)`.
```

Comment:
is this accomplished by setting the pod.spec.hostname field? is that valid for all combinations of $jobname and $index (e.g. the longest possible job name and a multi-digit index)?

Reply:
This is similar to statefulset, hostname is set to the pod name which is

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_utils.go#L88

Reply:
It seems that we don't validate that the job name is a valid DNS name like we do for statefulset: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/apis/apps/validation/validation.go#L45

Reply:
It is through the hostname field.

It doesn't look like we validate the length of the StatefulSet name to account for every possible index.
Should we do it just if the Job.spec.completionMode=Indexed (alpha in 1.21).

Reply:
Yes, I believe we can do this for the indexed completion mode only since it is still alpha, and by that we don't break job in non-indexed mode.

Reply:
lets continue the discussion in the other PR

Reply:
We should also check that it is a valid DNs name, not just the length. @johnbelamaric can we use [`IsDNS1123Label`](https://github.com/kubernetes/kubernetes/blob/31030820be979ea0b2c39e08eb18fddd71f353ed/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L187)?

Reply:
we should make sure that whatever construction we make from the name and biggest possible index will pass the same validation as the hostname field

I think that means 
`ValidateDNS1123Label(name + "-" + biggest possible index)`

also, if the biggest possible index can change, then that validation should be done on updates to indexed jobs that attempt to change the biggest possible index

Reply:
@ahg-g the name is already validated. The check is the same for every controller (including statefulset).

So you are proposing we do the validation dynamic, based on the `Job.spec.completions`. That works for me. It will give more room for Job name, as not all will use the max possible int32.

Today, the number of completions is immutable, but I'm thinking of changing that in the next release. So, let's check during Update.

Reply:
New validation LGTM. Note that the same validation function is also called during Update, so no need for any extra changes in the Update strategy.